### PR TITLE
Add several string functions to stdlib

### PR DESCRIPTION
**Pull request by nofun97**
Added:
1. `sub(replace, old, new)`
2. `split(str, separator)`
3. `lower(str)`
4. `upper(str)`
5. `title(str)`
6. `concat(list_of_str)`
7. `contains(str, substr)`
8. `join(list_of_str, separator)`
9. Simple `ArrayEnumerator` and `Array` interface
10. Corresponding test

### CODE REVIEW

File: rel/value_set_array.go [hunk 1]
```diff
@@ -1,7 +1,20 @@
 package rel
 
+import "github.com/arr-ai/frozen"
+
 const ArrayItemAttr = "@item"
 
+// Array represents a sequence of elements.
+type Array interface {
```
Comment (marcelocantos) [inline rel/value_set_array.go]: This interface serves no purpose. You can just define `ArrayEnumerator` as a free function.

Comment (nofun97) [reply rel/value_set_array.go]: Wouldn't it be possible in the future to have more functionalities in the `Array` interface? Maybe assigning at an index

Comment (marcelocantos) [reply rel/value_set_array.go]: A more useful addition would be an `Array` struct that implements array-like sets as Go slices. It would be very similar to `String`.

File: syntax/str_stdlib_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`["this is a test"]`,
+		`//.str.split("this is a test", ",")`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`["th", " ", " a test"]`,
+		`//.str.split("this is a test", "is")`,
+	)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: One-liners are much more readable:
```go
AssertCodesEvalToSameValue(t, `["this", "is", "a", "test"]`, `//.str.split("this is a test", " ")`)
AssertCodesEvalToSameValue(t, `["this is a test"]         `, `//.str.split("this is a test", ",")`)
AssertCodesEvalToSameValue(t, `["th", " ", " a test"]     `, `//.str.split("this is a test", "is")`)
```

I generally have a hard-limit of 120 columns. Within that, it's a judgment call.

File: syntax/str_stdlib_test.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`"this is still a test"`,
+		`//.str.sub("this is still a test", "doesn't matter", "hello there")`,
+	)
+	assertExprPanics(t, `//.str.sub("hello there", "test", 1)`)
+}
+
+func TestSplit(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Add some edge cases. Splitter not found, empty splitter, empty target.

File: syntax/str_stdlib_test.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`"this is a test"`,
+		`//.str.lower("ThIs is A TeST")`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"this is a test"`,
+		`//.str.lower("this is a test")`,
+	)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liners

File: syntax/str_stdlib_test.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`"THIS IS A TEST"`,
+		`//.str.upper("ThIs is A TeST")`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"THIS IS A TEST"`,
+		`//.str.upper("this is a test")`,
+	)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liners

File: syntax/str_stdlib_test.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`"ThIs Is A TeST"`,
+		`//.str.title("ThIs is A TeST")`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"This Is A Test"`,
+		`//.str.title("this is a test")`,
+	)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liners

File: syntax/str_stdlib_test.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`true`,
+		`//.str.contains("this is a test", "is a test")`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`false`,
+		`//.str.contains("this is a test", "is not a test")`,
+	)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liners

File: syntax/str_stdlib_test.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`"this is a test"`,
+		`//.str.concat(["this", " is", " a", " test"])`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"this"`,
+		`//.str.concat(["this"])`,
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liners

File: syntax/str_stdlib_test.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+	AssertCodesEvalToSameValue(
+		t,
+		`"this is a test"`,
+		`//.str.join(["this", "is", "a", "test"], " ")`,
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"this"`,
+		`//.str.join(["this"], ",")`,
+	)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liners

File: syntax/str_stdlib_test.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+	assertExprPanics(t, `//.str.join("this", 2)`)
+}
+
+func assertExprPanics(t *testing.T, code string) {
+	assert.Panics(t, func() {
+		AssertCodesEvalToSameValue(
+			t,
+			`"doesn't matter"`,
+			code,
+		)
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: one-liner

File: syntax/str_stdlib_test.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
+package syntax
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSub(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Add some edge cases. empty text, empty search string, empty replacement, search == text. replacement == search.

File: syntax/str_stdlib_test.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`["th", " ", " a test"]`,
+		`//.str.split("this is a test", "is")`,
+	)
+	assertExprPanics(t, `//.str.split("this is a test", 1)`)
+}
+
+func TestLower(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Test empty string, unicode, no text.

File: syntax/str_stdlib_test.go [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"this is a test"`,
+		`//.str.lower("this is a test")`,
+	)
+	assertExprPanics(t, `//.str.lower(123)`)
+}
+
+func TestUpper(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Test empty string, unicode, no text.

File: syntax/str_stdlib_test.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"THIS IS A TEST"`,
+		`//.str.upper("this is a test")`,
+	)
+	assertExprPanics(t, `//.str.upper(123)`)
+}
+
+func TestTitle(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Test empty string, unicode, no text, single word.

File: syntax/str_stdlib_test.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"This Is A Test"`,
+		`//.str.title("this is a test")`,
+	)
+	assertExprPanics(t, `//.str.title(123)`)
+}
+
+func TestContains(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Test empty text, empty search, empty both, text = search.

File: syntax/str_stdlib_test.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`false`,
+		`//.str.contains("this is a test", "is not a test")`,
+	)
+	assertExprPanics(t, `//.str.contains(123, 124)`)
+}
+
+func TestConcat(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Test some cases with empty strings. At the start, in the middle, at the end.

File: syntax/str_stdlib_test.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+	)
+	AssertCodesEvalToSameValue(
+		t,
+		`"this"`,
+		`//.str.concat(["this"])`,
+	)
+	assertExprPanics(t, `//.str.concat("this")`)
+}
+
+func TestJoin(t *testing.T) {
```
Comment (marcelocantos) [inline syntax/str_stdlib_test.go]: Test some cases with empty strings. Also test empty separator.
